In [1]:
import pandas as pd
import numpy as np

In [2]:
adult = pd.read_csv("adult.data", names = ["age", "work_class", "fnlwgt", "education", "education_num", "martial_status", "occupation", "relationship", "race", "sex","capital_gain", "capital-loss", "hours_per_week", "native_country", "label"],
                    index_col=False , skipinitialspace = True)
adult.head()
#adult.info()

age        work_class  fnlwgt  education  education_num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       martial_status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital_gain  capital-loss  hours_per_week native_country  label  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

TASK 1 : 
1. Count how many people for each category in the column "label")


In [3]:
adult.groupby("label").size()

#adult.groupby("label").count() Gives all values in the column based on <= 50k and > 50k 

label
<=50K    24720
>50K      7841
dtype: int64

So there are 24,720 people who earn less than or equal to 50k and 7841 who earn more than 50k 


Q2: 
For each combination of age and education, find how many people earn more than  50k 


In [4]:
#Created a new df that only includes dataset where label is >50k. Then grouped this dataframe 
adult_50k = adult.loc[adult["label"] == ">50K"]
adult_50k.groupby(["age", "education"])["label"].count()
#The following table counts based on age and education, how many people earn more than 50k0 

age  education   
19   12th            1
     7th-8th         1
21   Assoc-acdm      1
     HS-grad         1
     Some-college    1
                    ..
84   Some-college    1
90   Bachelors       2
     HS-grad         2
     Masters         3
     Prof-school     1
Name: label, Length: 572, dtype: int64

Q3. 
Pivot the education and for each occupation find people earning more than 50k for each gender 

In [5]:
#Pivot dataset:
grouped_50k = adult[adult["label"] == ">50K"]
grouped_50k.pivot_table("label", index = "occupation", columns = "sex", aggfunc = "size")


sex                Female    Male
occupation                       
?                    52.0   139.0
Adm-clerical        212.0   295.0
Armed-Forces          NaN     1.0
Craft-repair         20.0   909.0
Exec-managerial     280.0  1688.0
Farming-fishing       2.0   113.0
Handlers-cleaners     4.0    82.0
Machine-op-inspct    20.0   230.0
Other-service        51.0    86.0
Priv-house-serv       1.0     NaN
Prof-specialty      385.0  1474.0
Protective-serv      10.0   201.0
Sales                88.0   895.0
Tech-support         45.0   238.0
Transport-moving      9.0   311.0

q4:
Pivot using occupation as index and column as martial_status and sex. Count of those who earn more than 50k 

In [6]:
#Located the labels above 50k and then use those columns to pivot. 
#New method: 
adult.loc[adult["label"] == ">50K"].pivot_table("label", index = "occupation", columns = ["martial_status", "sex"], aggfunc = "count")

#ALTERNATIVE METHOD:
#grouped_50k.pivot_table("label", index = "occupation", columns = ["martial_status", "sex"], aggfunc = "count")

martial_status    Divorced       Married-AF-spouse      Married-civ-spouse  \
sex                 Female  Male            Female Male             Female   
occupation                                                                   
?                      4.0   2.0               NaN  NaN               42.0   
Adm-clerical          15.0  11.0               1.0  NaN              171.0   
Armed-Forces           NaN   NaN               NaN  NaN                NaN   
Craft-repair           7.0  38.0               1.0  1.0               10.0   
Exec-managerial       66.0  77.0               1.0  NaN              155.0   
Farming-fishing        1.0   4.0               NaN  1.0                NaN   
Handlers-cleaners      NaN   4.0               NaN  NaN                4.0   
Machine-op-inspct      1.0   3.0               NaN  NaN               17.0   
Other-service          6.0   5.0               2.0  NaN               32.0   
Priv-house-serv        NaN   NaN               NaN  NaN                NaN   
Prof-specialty        57.0  63.0               1.0  2.0              231.0   
Protective-serv        2.0  11.0               NaN  NaN                2.0   
Sales                 15.0  40.0               NaN  NaN               50.0   
Tech-support           4.0  13.0               NaN  NaN               33.0   
Transport-moving       1.0  13.0               NaN  NaN                7.0   

martial_status            Married-spouse-absent      Never-married         \
sex                  Male                Female Male        Female   Male   
occupation                                                                  
?                   129.0                   NaN  1.0           2.0    7.0   
Adm-clerical        264.0                   1.0  1.0          17.0   16.0   
Armed-Forces          1.0                   NaN  NaN           NaN    NaN   
Craft-repair        823.0                   NaN  4.0           2.0   33.0   
Exec-managerial    1511.0                   3.0  5.0          42.0   70.0   
Farming-fishing     101.0                   1.0  NaN           NaN    5.0   
Handlers-cleaners    75.0                   NaN  NaN           NaN    2.0   
Machine-op-inspct   222.0                   NaN  1.0           2.0    3.0   
Other-service        72.0                   1.0  1.0           5.0    7.0   
Priv-house-serv       NaN                   NaN  NaN           NaN    NaN   
Prof-specialty     1276.0                   3.0  8.0          72.0  105.0   
Protective-serv     179.0                   NaN  NaN           6.0    9.0   
Sales               790.0                   2.0  1.0          10.0   50.0   
Tech-support        211.0                   NaN  NaN           7.0   10.0   
Transport-moving    284.0                   NaN  1.0           1.0    8.0   

martial_status    Separated       Widowed        
sex                  Female  Male  Female  Male  
occupation                                       
?                       NaN   NaN     4.0   NaN  
Adm-clerical            1.0   2.0     6.0   1.0  
Armed-Forces            NaN   NaN     NaN   NaN  
Craft-repair            NaN   6.0     NaN   4.0  
Exec-managerial         3.0  16.0    10.0   9.0  
Farming-fishing         NaN   1.0     NaN   1.0  
Handlers-cleaners       NaN   NaN     NaN   1.0  
Machine-op-inspct       NaN   1.0     NaN   NaN  
Other-service           1.0   1.0     4.0   NaN  
Priv-house-serv         NaN   NaN     1.0   NaN  
Prof-specialty         11.0  11.0    10.0   9.0  
Protective-serv         NaN   1.0     NaN   1.0  
Sales                   1.0   4.0    10.0  10.0  
Tech-support            NaN   3.0     1.0   1.0  
Transport-moving        NaN   3.0     NaN   2.0

q5



In [7]:
grouped_50k.pivot_table("label", index = ["occupation", "sex"], columns = "martial_status", aggfunc = "count")

martial_status            Divorced  Married-AF-spouse  Married-civ-spouse  \
occupation        sex                                                       
?                 Female       4.0                NaN                42.0   
                  Male         2.0                NaN               129.0   
Adm-clerical      Female      15.0                1.0               171.0   
                  Male        11.0                NaN               264.0   
Armed-Forces      Male         NaN                NaN                 1.0   
Craft-repair      Female       7.0                1.0                10.0   
                  Male        38.0                1.0               823.0   
Exec-managerial   Female      66.0                1.0               155.0   
                  Male        77.0                NaN              1511.0   
Farming-fishing   Female       1.0                NaN                 NaN   
                  Male         4.0                1.0               101.0   
Handlers-cleaners Female       NaN                NaN                 4.0   
                  Male         4.0                NaN                75.0   
Machine-op-inspct Female       1.0                NaN                17.0   
                  Male         3.0                NaN               222.0   
Other-service     Female       6.0                2.0                32.0   
                  Male         5.0                NaN                72.0   
Priv-house-serv   Female       NaN                NaN                 NaN   
Prof-specialty    Female      57.0                1.0               231.0   
                  Male        63.0                2.0              1276.0   
Protective-serv   Female       2.0                NaN                 2.0   
                  Male        11.0                NaN               179.0   
Sales             Female      15.0                NaN                50.0   
                  Male        40.0                NaN               790.0   
Tech-support      Female       4.0                NaN                33.0   
                  Male        13.0                NaN               211.0   
Transport-moving  Female       1.0                NaN                 7.0   
                  Male        13.0                NaN               284.0   

martial_status            Married-spouse-absent  Never-married  Separated  \
occupation        sex                                                       
?                 Female                    NaN            2.0        NaN   
                  Male                      1.0            7.0        NaN   
Adm-clerical      Female                    1.0           17.0        1.0   
                  Male                      1.0           16.0        2.0   
Armed-Forces      Male                      NaN            NaN        NaN   
Craft-repair      Female                    NaN            2.0        NaN   
                  Male                      4.0           33.0        6.0   
Exec-managerial   Female                    3.0           42.0        3.0   
                  Male                      5.0           70.0       16.0   
Farming-fishing   Female                    1.0            NaN        NaN   
                  Male                      NaN            5.0        1.0   
Handlers-cleaners Female                    NaN            NaN        NaN   
                  Male                      NaN            2.0        NaN   
Machine-op-inspct Female                    NaN            2.0        NaN   
                  Male                      1.0            3.0        1.0   
Other-service     Female                    1.0            5.0        1.0   
                  Male                      1.0            7.0        1.0   
Priv-house-serv   Female                    NaN            NaN        NaN   
Prof-specialty    Female                    3.0           72.0       11.0   
                  Male                      8.0          105.0       11.

q6

In [8]:
#adult.head()
pd.crosstab(index = adult.education , columns = adult.sex)   #Using crosstab with index as education from adult and column as sex


sex           Female  Male
education                 
10th             295   638
11th             432   743
12th             144   289
1st-4th           46   122
5th-6th           84   249
7th-8th          160   486
9th              144   370
Assoc-acdm       421   646
Assoc-voc        500   882
Bachelors       1619  3736
Doctorate         86   327
HS-grad         3390  7111
Masters          536  1187
Preschool         16    35
Prof-school       92   484
Some-college    2806  4485

## Task 2

In [9]:
user = pd.read_csv("u.user.txt", sep = "|")  #Loading the dataset
user.head()

user_id  age gender  occupation zip_code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

q1


In [10]:
mean_age = user["age"].groupby(user["occupation"]).mean()
print(mean_age)

occupation
administrator    38.746835
artist           31.392857
doctor           43.571429
educator         42.010526
engineer         36.388060
entertainment    29.222222
executive        38.718750
healthcare       41.562500
homemaker        32.571429
lawyer           36.750000
librarian        40.000000
marketing        37.615385
none             26.555556
other            34.523810
programmer       33.121212
retired          63.071429
salesman         35.666667
scientist        35.548387
student          22.081633
technician       33.148148
writer           36.311111
Name: age, dtype: float64


q2. 
FInd female ratio per occupations and sort it from the most to least 

In [15]:
female_count = user.groupby("occupation")["gender"].value_counts()  #Counting number of females and males grouped by occupation.
# print(female_count)
total_gender = user.groupby("occupation")["gender"].size() #Counting total number of people in each occupation
percent = (female_count/total_gender)*100  #Calculating the ratio
#percent.head()
occupationbygender_ratio = percent.to_frame().unstack() #Need to convert the percent which is in series form to dataframe for further calculations
#unstack the percent ratio dataframe so that male and female now in different columns


occupationbygender_ratio = occupationbygender_ratio.droplevel(0, axis=1) #drop the index "gender" in position 1 through column to extract female occupation_gender ratio
occupation_female_ratio =  occupationbygender_ratio.drop("M", axis = 1 )
#occupation_female_ratio.head()

occupation_female_ratio = occupation_female_ratio.sort_values('F', ascending=False)
print(occupation_female_ratio)

#occupation_female_ratio.info()

gender                 F
occupation              
homemaker      85.714286
healthcare     68.750000
librarian      56.862745
artist         46.428571
administrator  45.569620
none           44.444444
writer         42.222222
marketing      38.461538
other          34.285714
student        30.612245
educator       27.368421
salesman       25.000000
lawyer         16.666667
entertainment  11.111111
scientist       9.677419
executive       9.375000
programmer      9.090909
retired         7.142857
technician      3.703704
engineer        2.985075
doctor               NaN


q3: For each occupation, calculate the min and max age

In [12]:
user.groupby("occupation")["age"].agg(["min", "max"])  #Groupby occupation and use the column age. Use function min and max

min  max
occupation             
administrator   21   70
artist          19   48
doctor          28   64
educator        23   63
engineer        22   70
entertainment   15   50
executive       22   69
healthcare      22   62
homemaker       20   50
lawyer          21   53
librarian       23   69
marketing       24   55
none            11   55
other           13   64
programmer      20   63
retired         51   73
salesman        18   66
scientist       23   55
student          7   42
technician      21   55
writer          18   60

q4: for combination of occupation and gender, calculate mean age

In [13]:
user.groupby(["occupation", "gender"])["age"].mean().unstack() #Groupby two categories, pass them as list and use column age to calculate the mean

gender                 F          M
occupation                         
administrator  40.638889  37.162791
artist         30.307692  32.333333
doctor               NaN  43.571429
educator       39.115385  43.101449
engineer       29.500000  36.600000
entertainment  31.000000  29.000000
executive      44.000000  38.172414
healthcare     39.818182  45.400000
homemaker      34.166667  23.000000
lawyer         39.500000  36.200000
librarian      40.000000  40.000000
marketing      37.200000  37.875000
none           36.500000  18.600000
other          35.472222  34.028986
programmer     32.166667  33.216667
retired        70.000000  62.538462
salesman       27.000000  38.555556
scientist      28.333333  36.321429
student        20.750000  22.669118
technician     38.000000  32.961538
writer         37.631579  35.346154